File: backend/schemas/reports.py
"""Схеми для звітних JSON-ендпоінтів (баланси виробів тощо)."""

from typing import List, Optional
from pydantic import BaseModel


class ProductBalanceClientRow(BaseModel):
    client_id: int
    client_name: str
    client_kind: str
    ordered_qty: float = 0       # з orders (для системних клієнтів завжди 0)
    invoiced_qty: float = 0
    invoiced_sum: float = 0
    other_qty: float = 0         # списання+пайок+інше, об'єднано (див. tags)
    other_sum: float = 0
    tags: List[str] = []         # 'exchange'/'stale'/'surplus'/'writeoff'/'ration'/'other'


class ProductBalanceProduct(BaseModel):
    product_id: int
    name: str
    short_name: Optional[str] = None
    # ── Report-parity (ідентично Секції 1 денного звіту) ──
    ordered_qty: float
    baked_qty: float
    baked_entered: bool
    exchange_qty: float
    shop_qty: float
    # ── Розширення дашборду ──
    invoiced_qty: float
    invoiced_sum: float
    writeoff_qty: float
    writeoff_sum: float
    ration_qty: float
    ration_sum: float
    other_qty: float
    other_sum: float
    diff_qty: Optional[float] = None
    clients: List[ProductBalanceClientRow] = []


class ProductBalanceCategory(BaseModel):
    category_id: int
    category_name: str
    ordered_qty: float
    baked_qty: float
    baked_entered: bool
    exchange_qty: float
    shop_qty: float
    invoiced_qty: float
    invoiced_sum: float
    writeoff_qty: float
    writeoff_sum: float
    ration_qty: float
    ration_sum: float
    other_qty: float
    other_sum: float
    diff_qty: Optional[float] = None
    products: List[ProductBalanceProduct] = []


class ProductBalancesOut(BaseModel):
    date: str
    categories: List[ProductBalanceCategory] = []

File: frontend/src/api/reports.ts
import { api } from './client'

export interface ProductBalanceClientRow {
  client_id: number
  client_name: string
  client_kind: string
  ordered_qty: number
  invoiced_qty: number
  invoiced_sum: number
  other_qty: number           // списання+пайок+інше, об'єднано (див. tags)
  other_sum: number
  tags: string[]               // 'exchange'/'stale'/'surplus'/'writeoff'/'ration'/'other'
}

export interface ProductBalanceProduct {
  product_id: number
  name: string
  short_name: string | null
  // report-parity (ідентично Секції 1 денного звіту)
  ordered_qty: number
  baked_qty: number
  baked_entered: boolean
  exchange_qty: number
  shop_qty: number
  // розширення дашборду
  invoiced_qty: number
  invoiced_sum: number
  writeoff_qty: number
  writeoff_sum: number
  ration_qty: number
  ration_sum: number
  other_qty: number
  other_sum: number
  diff_qty: number | null
  clients: ProductBalanceClientRow[]
}

export interface ProductBalanceCategory {
  category_id: number
  category_name: string
  ordered_qty: number
  baked_qty: number
  baked_entered: boolean
  exchange_qty: number
  shop_qty: number
  invoiced_qty: number
  invoiced_sum: number
  writeoff_qty: number
  writeoff_sum: number
  ration_qty: number
  ration_sum: number
  other_qty: number
  other_sum: number
  diff_qty: number | null
  products: ProductBalanceProduct[]
}

export interface ProductBalancesOut {
  date: string
  categories: ProductBalanceCategory[]
}

export function fetchProductBalances(date: string): Promise<ProductBalancesOut> {
  return api.get<ProductBalancesOut>(`/reports/product-balances?date=${date}`)
}
